File: java/src/com/jambit/practicalresult/Services.java
package com.jambit.practicalresult;

import io.vavr.control.Either;

import java.util.function.Function;

public class Services {

    interface Service<T> extends Function<T, Either<String, T>> {
    }

    public static Service<Integer> byTwo = (i) -> Either.right(i * 2);

    public static Service<Integer> plusOneIfEven =
            (i) -> i % 2 != 0 ? Either.left("I don't like odds!")
                              : Either.right(i + 1);
}

File: typescript/src/services.ts
import {failure, Result, success} from "./result";

export type MapService<T> = (t: T) => Result<String, T>

export const byTwo: MapService<number> = i => success(i * 2)

export const plusOneIfEven: MapService<number> = i => i % 2 != 0
    ? failure("I don't like odds!")
    : success(i + 1)
